docs/developers/pushdata.md

```diff
@@ -0,0 +1,113 @@
+---
+license: |
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+  http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+
+
+# Push Data
+
+This articles describes the detailed design of the process of push data.
+
+## API specification
+The push data API is as follows:
+```shell
+  public abstract int pushData(
+      int shuffleId,
+      int mapId,
+      int attemptId,
+      int partitionId,
+      byte[] data,
+      int offset,
+      int length,
+      int numMappers,
+      int numPartitions)
+      throws IOException;
+```
+
+- `shuffleId` is the unique shuffle id of the application
+- `mapId` is the map id of the shuffle
+- `attemptId` is the attempt id of the map task, i.e. speculative task or task rerun for Apache Spark
+- `partitionId` is the partition id the data belongs to
+- `data`,`offset`,`length` specifies the bytes to be pushed
+- `numMappers` is the number map tasks in the shuffle
+- `numPartitions` is the number of partitions in the shuffle
+
+## Lazy Shuffle Register
+The first time `pushData` is called, Client will check whether the shuffle id has been registered. If Not,
+it sends `RegisterShuffle` to `LifecycleManager`, `LifecycleManager` then sends `RequestSlots` to `Master`.
+`RequestSlots` specifies how many `PartitionLocation`s this shuffle requires, each `PartitionLocation` logically
+responds to data of some partition id.
+
+Upon receiving `RequestSlots`, `Master` allocates slots for the shuffle among `Worker`s. If replication is turned on,
+`Master` allocates a pair of `Worker`s for each `PartitionLocation` to store two replicas for each `PartitionLocation`.
+The detailed allocation strategy can be found in [Slots Allocation](/developers/slotsallocation). `Master` then
```

Comment:
I'm not sure if absolute path work since the docs site is not start from root `/`, may be relative path works